src/main/java/io/supertokens/inmemorydb/queries/TOTPQueries.java

```diff
@@ -0,0 +1,197 @@
+            pst.setString(3, oldDeviceName);
+        });
+    }
+
+    public static TOTPDevice[] getDevices(Start start, String userId)
+            throws StorageQueryException, SQLException {
+        String QUERY = "SELECT * FROM " + Config.getConfig(start).getTotpUserDevicesTable()
+                + " WHERE user_id = ?;";
+
+        return execute(start, QUERY, pst -> pst.setString(1, userId), result -> {
+            List<TOTPDevice> devices = new ArrayList<>();
+            while (result.next()) {
+                devices.add(TOTPDeviceRowMapper.getInstance().map(result));
+            }
+
+            return devices.toArray(TOTPDevice[]::new);
+        });
+    }
+
+    public static int insertUsedCode(Start start, TOTPUsedCode code) throws SQLException, StorageQueryException {
+        String QUERY = "INSERT INTO " + Config.getConfig(start).getTotpUsedCodesTable()
+                + " (user_id, device_name, code, is_valid_code, expiry_time_ms) VALUES (?, ?, ?, ?, ?);";
+
+        return update(start, QUERY, pst -> {
+            pst.setString(1, code.userId);
+            pst.setString(2, code.deviceName);
+            pst.setString(3, code.code);
+            pst.setBoolean(4, code.isValidCode);
+            pst.setLong(5, code.expiryTime);
+        });
+    }
+
+    public static TOTPUsedCode[] getUsedCodes(Start start, String userId) throws SQLException, StorageQueryException {
+        String QUERY = "SELECT * FROM " +
+                Config.getConfig(start).getTotpUsedCodesTable()
+                + " WHERE user_id = ? AND expiry_time_ms > ?;";
+        return execute(start, QUERY, pst -> {
+            pst.setString(1, userId);
+            pst.setLong(2, System.currentTimeMillis());
+        }, result -> {
+            List<TOTPUsedCode> codes = new ArrayList<>();
+            while (result.next()) {
+                codes.add(TOTPUsedCodeRowMapper.getInstance().map(result));
+            }
+
+            return codes.toArray(TOTPUsedCode[]::new);
+        });
+    }
+
+    public static int removeExpiredCodes(Start start)
+            throws StorageQueryException, SQLException {
+        String QUERY = "DELETE FROM " + Config.getConfig(start).getTotpUsedCodesTable()
+                + " WHERE expiry_time_ms < ?;";
+
+        return update(start, QUERY, pst -> pst.setLong(1, System.currentTimeMillis()));
+    }
+
+    public static int removeUsedCodes(Start start, String userId, String deviceName)
+            throws StorageQueryException, SQLException {
+        // Remove codes where userId matches the given userId
+        // ONLY required for inmemorydb, as it does not support foreign key constraints.
+        String QUERY = "DELETE FROM " + Config.getConfig(start).getTotpUsedCodesTable()
+                + " WHERE user_id = ? AND device_name = ?;";
+
+        return update(start, QUERY, pst -> {
+            pst.setString(1, userId);
+            pst.setString(2, deviceName);
+        });
+    }
```

Comment:
remove this since this is not a plugin interface function directly.

Reply:
Removed.